src/src/Controller/CustomerSyncController.php

```diff
@@ -0,0 +1,266 @@
+
+        $subdomain = $configuration->get('SMAILY_SUBDOMAIN');
+        $username = $configuration->get('SMAILY_USERNAME');
+        $password = $configuration->get('SMAILY_PASSWORD');
+
+        if (!empty($subdomain) && !empty($username) && !empty($password)) {
+            $this->api = new Api($subdomain, $username, $password);
+        }
+    }
+
+    public function sync(string $token): bool
+    {
+        if ($token !== $this->configuration->get('SMAILY_CUSTOMER_CRON_TOKEN')) {
+            echo 'Access denied!';
+
+            return false;
+        }
+
+        if (!$this->configuration->get('SMAILY_ENABLE_CUSTOMER_SYNC')) {
+            echo 'User synchronization disabled!';
+
+            return false;
+        }
+
+        if (empty($this->api)) {
+            echo 'Account setup not finished!';
+
+            return false;
+        }
+
+        $unsubscribers_synchronized = $this->removeUnsubscribers(self::UNSUBSCRIBERS_BATCH_LIMIT);
+        if (!$unsubscribers_synchronized) {
+            echo 'Failed to synchronize unsubscribers from Smaily!';
+
+            Logger::logMessageWithSeverity('Customer sync failed - unsubscribers are not removed', 1);
+
+            return false;
+        }
+
+        // Don't sync customers if failed to remove unsubscribers.
+        $subscribers_synchronized = $this->sendSubscribersToSmaily(self::SUBSCRIBERS_BATCH_LIMIT);
+        if (!$subscribers_synchronized) {
+            echo 'Failed to synchronize subscribers from the store!';
+
+            Logger::logMessageWithSeverity('Customer sync failed - failed to send subscribers to Smaily', 1);
+
+            return false;
+        }
+
+        echo 'Subscribers synchronization finished!';
+
+        return true;
+    }
+
+    /**
+     * Get unsubscribers from Smaily and change the subscription status to unsubscribed in store.
+     *
+     * @param int $limit limit request size
+     *
+     * @return bool success status
+     */
+    private function removeUnsubscribers(int $limit = 1000)
+    {
+        $offset = 0;
+
+        while (true) {
+            $response = $this->api->listUnsubscribers($limit, $offset);
+
+            // Stop if error.
+            if ($response->getStatusCode() !== 200) {
+                Logger::logErrorWithFormatting('Failed fetching unsubscribers.');
+
+                return false;
+            }
+
+            $body = json_decode($response->getBody()->getContents(), true);
+
+            // Stop if no more subscribers.
+            if (empty($body)) {
+                break;
+            }
+
+            // Remove subscribed status for unsubscribers.
+            $subscriberCollectionModel = new SubscriberCollection();
+            $result = $subscriberCollectionModel->batchUnsubscribeByEmail(
+                array_map(
+                    function ($item) {
+                        return $item['email'];
+                    },
+                    $body,
+                )
```

Comment:
PHP has a handy function for this: [`array_column`](https://www.php.net/manual/en/function.array-column.php).

Reply:
Nice https://github.com/sendsmaily/smaily-prestashop-module/pull/57/commits/4200394aaa2dc66be01e3d940e2991b22e56b6a2